### PR TITLE
feat: Color scheme customization

### PR DESCRIPTION
**Pull request by matthewlipski**
This PR overhauls how users can style the BlockNote editor. This is done in 3 ways:

- Allowing users to change the colors, fonts, and border radii of the default UI (includes editor itself, and all menus/toolbars)
- Allowing users to add HTML class names to elements within the editor.
- Overriding the CSS for UI components.

There are still a few points of contention:

One of the main limitations is the extended theming customizability. I think most people are interested in just changing colors and fonts, for which we have a good solution, but overriding CSS could probably use some improvements for the future.

Another limitation is that the HTML classes set on elements cannot be dynamically changed atm. I can see this being a blocker for some as setting `hidden`, `locked`, or `selected` classes is a pretty common use case. However, the implementation is actually pretty complex so not sure it's worth our time atm.

### CODE REVIEW

File: examples/editor/src/App.tsx [hunk 1]
```diff
@@ -1,26 +1,150 @@
...
-    editorDOMAttributes: {
-      class: styles.editor,
-      "data-test": "editor",
+    blockSchema: schema,
+    slashMenuItems: [
+      ...getDefaultReactSlashMenuItems(schema),
+      insertImage,
+      insertReactImage,
+    ],
+    domAttributes: {
```
Comment (YousefED) [inline examples/editor/src/App.tsx:118]: should we be able to pass custom domattributes for blocks here?

File: packages/core/src/extensions/Blocks/api/block.ts [hunk 2]
```diff
@@ -151,8 +151,20 @@ export function createBlockSpec<
       return ({ HTMLAttributes, getPos }) => {
         // Create blockContent element
         const blockContent = document.createElement("div");
-        // Sets blockContent class
-        blockContent.className = styles.blockContent;
+        // Add custom HTML attributes
+        const blockContentDOMAttributes =
```
Comment (YousefED) [inline packages/core/src/extensions/Blocks/api/block.ts:155]: why not blockContentDOMAttributes.blockContent.class?

File: packages/core/src/extensions/Blocks/nodes/BlockContainer.ts [hunk 3]
```diff
@@ -89,11 +86,16 @@ export const BlockContainer = Node.create<IBlock>({
...
         "div",
-        mergeAttributes(HTMLAttributes, {
-          // TODO: maybe remove html attributes from inner block
-          class: styles.block,
-          "data-node-type": this.name,
-        }),
+        mergeAttributes(
+          {
+            ...domAttributes,
+            class: domAttributes.class
```
Comment (YousefED) [inline packages/core/src/extensions/Blocks/nodes/BlockContainer.ts:92]: let's start using either https://www.npmjs.com/package/classnames for concatenation, or create a simple helper function ourselves

File: examples/editor/src/App.tsx [hunk 4]
```diff
@@ -1,26 +1,150 @@
...
+import { BlockSchema, defaultBlockSchema } from "@blocknote/core";
+import { Image, insertImage } from "../../../tests/utils/customblocks/Image";
+import {
+  insertReactImage,
+  ReactImage,
+} from "../../../tests/utils/customblocks/ReactImage";
 
 type WindowWithProseMirror = Window & typeof globalThis & { ProseMirror: any };
 
+const redTheme: Partial<Theme> = {
```
Comment (YousefED) [inline examples/editor/src/App.tsx:20]: I think adding dark and light modes as part of a theme creates a more complex API, and also more work (if you're creating a "red" theme, you'll need to implement both dark and light mode for it. In practice, what I think will happen, is that people will just insert duplicates or dummy modes for one).

I think what we should do instead is making BlockNoteLightTheme and BlockNoteDarkTheme default themes that we expose.

The `theme` prop on the editor could then take: `Theme | { dark: Theme, light: Theme}` and defaults to `{dark: BlockNoteDarkTheme, light: BlockNoteLightTheme}`. 

If a dark/light config is passed, we should auto-switch to the right theme based on browser settings

File: packages/react/src/BlockNoteTheme.ts [hunk 5]
```diff
@@ -1,82 +1,110 @@
...
+export type Theme = {
+  colors: ColorScheme;
+  borderRadius: number;
+  fontFamily: string;
+};
 
-  const sideMenu = useDarkTheme
-    ? blockNoteColorScheme[4]
-    : blockNoteColorScheme[2];
+export const blockNoteToMantineTheme = (
```
Comment (YousefED) [inline packages/react/src/BlockNoteTheme.ts:43]: My suggestion for dark / light would also make the code in this method a lot easier

File: packages/react/src/BlockNoteTheme.ts [hunk 6]
```diff
@@ -85,33 +113,69 @@ export const getBlockNoteTheme = (
...
+          : fullTheme.colors.highlightColors.blue.background.light,
+        useDarkTheme
+          ? fullTheme.colors.highlightColors.purple.background.dark
+          : fullTheme.colors.highlightColors.purple.background.light,
+        useDarkTheme
+          ? fullTheme.colors.highlightColors.pink.background.dark
+          : fullTheme.colors.highlightColors.pink.background.light,
       ],
     },
     components: {
```
Comment (YousefED) [inline packages/react/src/BlockNoteTheme.ts:181]: I'm wondering if we should make it easy to customize component styles as well for consumers.

I realize this will be mantine specific, so it should be an advanced option for sure. I also think we should only do it if it's easy to implement. The benefit I see is that it will basically allow almost full styleability of UI components (without creating your own) - even though users would probably need to dive into the source / mantine to see what exactly is stylable.

I think the main thing we need to know for this is whether we can extend from a mantine theme and keep both our `components` settings, and add new ones? i.e.: is the `components` setting extendable?

File: packages/react/src/BlockNoteView.tsx [hunk 7]
```diff
@@ -1,33 +1,66 @@
...
 export function BlockNoteView<BSchema extends BlockSchema>(
   props: {
     editor: BlockNoteEditor<BSchema>;
+    useDarkTheme?: boolean;
+    customTheme?: Partial<Theme>;
     children?: ReactNode;
   } & HTMLAttributes<HTMLDivElement>
 ) {
-  const { editor, children, ...rest } = props;
+  const { classes } = createStyles({ root: {} })(undefined, {
```
Comment (YousefED) [inline packages/react/src/BlockNoteView.tsx:50]: I think you can also pass the theme to `createStyles`? In that case the extra `UnThemedBlockNoteView` is not necessary

File: packages/react/src/SharedComponents/ColorPicker/components/ColorIcon.tsx [hunk 8]
```diff
@@ -7,21 +7,27 @@ export const ColorIcon = (
...
 
   return (
     <Box
+      className={classes.root}
       sx={(theme) => {
         return {
           backgroundColor:
             theme.colors.backgroundColors[
-              theme.other.colors.indexOf(backgroundColor)
+              theme.other.highlightColorNames.indexOf(backgroundColor)
```
Comment (YousefED) [inline packages/react/src/SharedComponents/ColorPicker/components/ColorIcon.tsx:25]: should `highlightColorNames` be part of the mantine theme? it seems unrelated tbh

**Review by YousefED — COMMENTED** [review]
Looks neat! Added some points to discuss, but I think we can quickly move to docs (keep it simple) & tests

File: packages/react/src/SharedComponents/ColorPicker/components/ColorIcon.tsx (hunk 8 above)
Comment (matthewlipski) [reply packages/react/src/SharedComponents/ColorPicker/components/ColorIcon.tsx:25]: I think it's important since they should really be different for the light and dark scheme, otherwise you get contrast issues and it generally looks bad. Maybe values for light & dark should just be hardcoded though?

File: examples/editor/src/App.tsx (hunk 1 above)
Comment (matthewlipski) [reply examples/editor/src/App.tsx:118]: Don't see why not, I think most people will still just add classes but imo it adds very little complexity so I think we may as well give people the option.

File: examples/editor/src/App.tsx (hunk 4 above)
Comment (matthewlipski) [reply examples/editor/src/App.tsx:20]: Yep I agree with this. How about the `useDarkMode` prop, do you think we should still keep it and have it default to the user's browser/OS setting? Or remove it entirely and just decide which to use based on browser/OS setting?

File: packages/react/src/BlockNoteTheme.ts (hunk 6 above)
Comment (matthewlipski) [reply packages/react/src/BlockNoteTheme.ts:181]: I think it should be possible, though I don't think we should spend much time on it. You can technically already do this with just CSS, e.g.:
```css
.mantine-Menu-item {
  background-color: blue !important;
}
```
So it's only really a semantic difference for a use case that we don't expect to be too common.

File: examples/editor/src/App.tsx (hunk 4 above)
Comment (YousefED) [reply examples/editor/src/App.tsx:20]: I think we can remove it. If someone wants to manually control it, they can just pass the light or darktheme directly as “theme”

File: packages/core/src/extensions/Blocks/api/block.ts (hunk 2 above)
Comment (matthewlipski) [reply packages/core/src/extensions/Blocks/api/block.ts:155]: For each of the DOM element types, you can add custom attributes, not just a custom class

File: packages/react/src/BlockNoteView.tsx (hunk 7 above)
Comment (matthewlipski) [reply packages/react/src/BlockNoteView.tsx:54]: Hmm don't think so - instead of an object containing styles (in this case `{ root: {} }`), you can provide a function which takes a theme and returns an object containing styles. But that theme param is still taken from the Context, I don't think there's a way to actually pass a theme in yourself:

```ts
export declare function createStyles<
  Key extends string = string, 
  Params = void, 
  Input extends Record<Key, CSSObject> = Record<Key, CSSObject>
>(
  input: 
    | (theme: MantineTheme, params: Params, createRef: (refName: string) => string) => Input
    | Input
): (
  params: Params, 
  options?: UseStylesOptions<Key>
) => {
  classes: { [key in keyof Input]: string; };
  cx: (...args: any) => string;
  theme: MantineTheme;
};
```

File: examples/editor/src/App.tsx [hunk 9]
```diff
@@ -1,26 +1,162 @@
...
+} satisfies Partial<Theme>;
+
+const schema = {
+  ...defaultBlockSchema,
+  image: Image,
+  reactImage: ReactImage,
+} satisfies BlockSchema;
+
+function App() {
+  const [theme, setTheme] = useState<
```
Comment (YousefED) [inline examples/editor/src/App.tsx:79]: This should note be here right? if the consumer passes dark + light to the theme prop, blocknote should do the selection?

File: packages/core/src/extensions/Blocks/api/block.ts (hunk 2 above)
Comment (YousefED) [reply packages/core/src/extensions/Blocks/api/block.ts:156]: Ok, you could do this:

constr { class, ...otherAttributes} = this.options.domAttributes?.blockContent || {};

File: packages/react/src/BlockNoteTheme.ts [hunk 10]
```diff
@@ -1,117 +1,160 @@
...
+    dark: darkDefaultTheme,
+  },
+  componentStyles?: BlockNoteComponentStyles
 ): MantineThemeOverride => {
-  const boxShadow = `0px 4px 8px ${
-    useDarkTheme ? blockNoteColorScheme[8] : blockNoteColorScheme[2]
-  }, 0px 0px 1px ${
-    useDarkTheme ? blockNoteColorScheme[6] : blockNoteColorScheme[1]
-  }`;
+  const fullTheme: Theme = shouldUseDarkTheme(theme)
```
Comment (YousefED) [inline packages/react/src/BlockNoteTheme.ts:87]: This code and `shouldUseDarkTheme` is difficult to read. To fix: first check if { light:..., dark: ... } was passed, or whether a single theme was passed.

i.e.:

```
if ('light' in theme && 'dark' in theme) {
    theme = browserWantsDarkTheme() ? theme.dark : theme.light;
}
```

File: packages/react/src/BlockNoteTheme.ts [hunk 11]
```diff
@@ -1,117 +1,160 @@
...
+        ...darkDefaultTheme,
+        ...("dark" in theme ? theme.dark : theme),
+      }
+    : {
+        ...lightDefaultTheme,
+        ...("light" in theme ? theme.light : theme),
+      };
+  console.log(`.${blockStyles.isEmpty} .${blockStyles.inlineContent}`);
+
+  const editorText = fullTheme.colors.editor.text;
```
Comment (YousefED) [inline packages/react/src/BlockNoteTheme.ts:98]: do we need line 98 - 114?

File: packages/react/src/BlockNoteTheme.ts [hunk 12]
```diff
@@ -246,22 +325,24 @@ export const getBlockNoteTheme = (
                 backgroundColor: hoveredBackground,
               },
             },
+            ...componentStyles?.Toolbar,
```
Comment (YousefED) [inline packages/react/src/BlockNoteTheme.ts:328]: The difficulty now of course is that if I want to change sth to `.mantine-UnstyledButton-root`, it will override the stuff from line 290. I suppose there's no easy way to prevent this? (again, let's not spend too much time on this)

File: packages/react/src/ReactBlockSpec.tsx [hunk 13]
```diff
@@ -38,16 +48,36 @@ export type ReactBlockConfig<
...
-    className={`${props.className ? props.className + " " : ""}${
-      blockStyles.inlineContent
-    }`}
-  />
-);
+) => {
+  const inlineContentDOMAttributes =
+    useContext(BlockNoteDOMAttributesContext).inlineContent || {};
+
+  const classNames = useMemo(
```
Comment (YousefED) [inline packages/react/src/ReactBlockSpec.tsx:59]: I think a usememo is overkill here as it's just a simple string concatenation

File: packages/react/src/defaultTheme.ts [hunk 14]
```diff
@@ -0,0 +1,124 @@
+import { Theme } from "./index";
```
Comment (YousefED) [inline packages/react/src/defaultTheme.ts:1]: nice that this file is extracted! super clear. Maybe rename to defaultThemes or create separate files for light / dark

File: packages/react/src/SharedComponents/ColorPicker/components/ColorIcon.tsx (hunk 8 above)
Comment (YousefED) [reply packages/react/src/SharedComponents/ColorPicker/components/ColorIcon.tsx:25]: I understand they should be themeable, I just don't understand the entire `indexOf` construction. Aren't we overcomplicating things?

File: examples/editor/src/App.tsx (hunk 9 above)
Comment (matthewlipski) [reply examples/editor/src/App.tsx:79]: It does, but it needs to also trigger a re-render if the browser theme changes, though this should probably be in `BlockNoteView` either way.

File: packages/react/src/BlockNoteTheme.ts (hunk 12 above)
Comment (matthewlipski) [reply packages/react/src/BlockNoteTheme.ts:328]: We would have to deep merge the existing styles with the new styles, I think theres a function from lodash actually which might do that.

File: packages/core/src/extensions/Blocks/api/block.ts (hunk 2 above)
Comment (matthewlipski) [reply packages/core/src/extensions/Blocks/api/block.ts:156]: Actually this also doesn't really work since `class` is a reserved word for class definitions

File: packages/react/src/BlockNoteTheme.ts [hunk 15]
```diff
@@ -1,316 +1,385 @@
...
+    purple: CombinedColor;
+    pink: CombinedColor;
+  };
+};
 
-  const border = `1px solid ${
-    useDarkTheme ? blockNoteColorScheme[7] : blockNoteColorScheme[1]
-  }`;
+export type Theme = {
+  type: "light" | "dark";
```
Comment (YousefED) [inline packages/react/src/BlockNoteTheme.ts:34]: why should a theme have a type?

File: packages/website/docs/docs/theming.md [hunk 16]
```diff
@@ -0,0 +1,261 @@
...
+
+`blockContainer:` The main container element for blocks. Contains both the block's content and its nested blocks.
+
+`blockGroup:` The wrapper element for all top-level blocks in the editor and nested blocks.
+
+`blockContent:` The wrapper element for a block's content.
+
+`inlineContent:` The wrapper element for a block's rich-text content.
+
+## Overriding CSS
```
Comment (YousefED) [inline packages/website/docs/docs/theming.md:176]: maybe add "(advanced)" to this

File: packages/react/src/BlockNoteView.tsx [hunk 17]
```diff
@@ -1,33 +1,136 @@
...
 import { FormattingToolbarPositioner } from "./FormattingToolbar/components/FormattingToolbarPositioner";
 import { HyperlinkToolbarPositioner } from "./HyperlinkToolbar/components/HyperlinkToolbarPositioner";
 import { SideMenuPositioner } from "./SideMenu/components/SideMenuPositioner";
 import { SlashMenuPositioner } from "./SlashMenu/components/SlashMenuPositioner";
+import { darkDefaultTheme, lightDefaultTheme } from "./defaultThemes";
+
+// Helper function which takes a partial theme provided by the user, and turns
+// it into a full theme. Either the light or dark theme is used, depending on
+// which ones the user provided and the browser settings.
+const getFullTheme = (baseTheme: "light" | "dark", theme?: PartialTheme) => {
```
Comment (YousefED) [inline packages/react/src/BlockNoteView.tsx:20]: I still don't really get this, let's discuss on discord

**Review by YousefED — COMMENTED** [review]
some more comments

File: packages/website/docs/docs/theming.md [hunk 18]
```diff
@@ -0,0 +1,261 @@
...
+export default function App() {
+  // Creates a new editor instance.
+  const editor: BlockNoteEditor = useBlockNote();
+
+  // Renders the editor instance using a React component.
+  return (
+    <BlockNoteView
+      editor={editor}
+      theme={"{{ getTheme(isDark) }}"}
+      componentStyles={(theme) => ({
```
Comment (YousefED) [inline packages/website/docs/docs/theming.md:198]: Should we make this part of the theme instead of a separate prop?

File: packages/website/docs/docs/theming.md [hunk 19]
```diff
@@ -0,0 +1,261 @@
...
+
+BlockNote allows you to change how the editor UI looks. You can change the theme of the default UI, or override its CSS styles.
+
+## Theming
+
+If you want to make small adjustments to BlockNote's UI, you can create a custom theme, which will let you change colors, fonts, and border radii. You can do this by passing partial `Theme` objects to `BlockNoteView`.
+
+Take a look at how this is done in the demo below:
+
+::: sandbox {template=react-ts}
```
Comment (YousefED) [inline packages/website/docs/docs/theming.md:25]: For sake of simplicity, I would just make one "red theme" and not a red light / dark theme

File: packages/core/src/extensions/Blocks/api/block.ts (hunk 2 above)
Comment (YousefED) [reply packages/core/src/extensions/Blocks/api/block.ts:156]: Check. We could decide to rename `class` to `className`, that's also inline with how React does this

File: packages/react/src/BlockNoteView.tsx (hunk 7 above)
Comment (YousefED) [reply packages/react/src/BlockNoteView.tsx]: I think you're right. In that case I have a small preference to rename `UnThemedBlockNoteView` to something else (`InnerBlockNoteView`)? While I get what you mean, "UnThemed" sounds to me like it's not themed, which is not really the case

File: packages/core/src/extensions/Blocks/api/block.ts (hunk 2 above)
Comment (matthewlipski) [reply packages/core/src/extensions/Blocks/api/block.ts:156]: Not sure if it's a good idea tbh, since we're not only applying the class to React components. Basic nodes like `blockContainer`, `blockGroup`, etc, are defined outside of React, so I think it's better to keep HTML naming conventions for this rather than React ones

File: packages/website/docs/docs/theming.md (hunk 19 above)
Comment (matthewlipski) [reply packages/website/docs/docs/theming.md:25]: I think it's nice to have both since it both shows off the auto switching from browser settings and saves us having to explain how to pass in a light & dark theme instead of just a single theme. Theming is also not really complicated compared to a lot of the other stuff, since 90% of it is defining colors, so imo we shouldn't run into the issue of people getting overloaded with information.

File: packages/react/src/BlockNoteView.tsx [hunk 20]
```diff
@@ -1,33 +1,97 @@
...
     children?: ReactNode;
   } & HTMLAttributes<HTMLDivElement>
 ) {
-  const { editor, children, ...rest } = props;
+  const {
+    theme = { light: lightDefaultTheme, dark: darkDefaultTheme },
+    ...rest
+  } = props;
+
+  const [browserTheme, setBrowserTheme] = useState<"light" | "dark">(
```
Comment (YousefED) [inline packages/react/src/BlockNoteView.tsx:61]: maybe use a separate hook like https://github.com/rfoel/use-prefers-color-scheme/blob/main/src/index.ts to isolate this logic?

File: packages/website/docs/docs/theming.md [hunk 21]
```diff
@@ -0,0 +1,259 @@
...
+  Theme,
+  useBlockNote,
+} from "@blocknote/react";
+import "@blocknote/core/style.css";
+
+// Default dark theme with additional component styles.
+const theme = {
+  ...darkDefaultTheme,
+  componentStyles: (theme) => ({
+    // Adds basic styling to the editor.
```
Comment (YousefED) [inline packages/website/docs/docs/theming.md:198]: You're also setting the font to comic sans, but don't explain that anywhere

**Review by YousefED — APPROVED** [review]
Looks great!
